code/parse/sexp.cpp

```diff
@@ -14742,6 +14756,73 @@ void sexp_send_message(int n)
 	sexp_send_one_message( name, who_from, priority, 0, 0, -1, do_hash_fallback );
 }
 
+ship* get_builtin_message_sender(const char* name) {
+	auto ship_entry = ship_registry_get(name);
+	if (ship_entry && ship_entry->has_shipp()) {
+		return ship_entry->shipp();
+	}
+	
+	auto wing_index = wing_lookup(name);
+	if (wing_index >= 0) {
+		auto ship_index = ship_get_random_ship_in_wing(wing_index, SHIP_GET_UNSILENCED);
+		return ship_index < 0 ? nullptr : &Ships[ship_index];
+	}
+
+	if (!stricmp(name, "<any wingman>")) {
+		auto ship_index = ship_get_random_player_wing_ship(SHIP_GET_UNSILENCED);
+		return ship_index < 0 ? nullptr : &Ships[ship_index];
+	}
+
+	return nullptr;
+}
+
+void sexp_send_builtin_message(int n) {
+	// Argument 1: Message type
+	auto type = get_builtin_message_type(CTEXT(n));
+	n = CDR(n);
+	// Argument 2: Subject (or <none>)
+	auto subject_name = CTEXT(n);
+	auto need_subject = stricmp(subject_name, SEXP_NONE_STRING);
+	auto subject_entry = ship_registry_get(subject_name);
+	auto have_subject = subject_entry && subject_entry->has_shipp();
+	n = CDR(n);
+	// Argument 3: Shuffle the senders?
+	auto shuffle = is_sexp_true(n);
+	n = CDR(n);
+	// Argument 4+: Possible senders
+	SCP_vector<const char*> sender_names;
+	while (n >= 0) {
+		sender_names.push_back(CTEXT(n));
+		n = CDR(n);
+	}
+
+	// Done consuming arguments; maybe short-circuit?
+	if (type == MESSAGE_NONE) {
+		return;
+	}
+	if (need_subject && !have_subject) {
+		return;
+	}
+
+	if (shuffle) {
+		std::random_shuffle(sender_names.begin(), sender_names.end());
+	}
+
+	auto subject = have_subject ? subject_entry->shipp() : nullptr;
+	for (auto sender_name : sender_names) {
+		if (is_special_sender(sender_name)) {
+			message_send_builtin(type, nullptr, subject);
+			return;
+		} else {
+			auto sender = get_builtin_message_sender(sender_name);
+			if (sender != nullptr) {
+				message_send_builtin(type, sender, subject);
+				return;
+			}
+		}
+	}
```

Comment:
Dunno about the use case, so this can be ignored if you think it doesn't matter, but:
Does it maybe make sense to return whether a message was actually sent? If all given senders are invalid or silenced, it could be important to tell the fredder.

Reply:
I thought about this, but I don't think it's a good idea. We'd need to make a second operator that differs only in return type: just making it return a boolean would force the FREDer to wrap it in a `when` in the common case, and adding an optional argument for a variable that holds the result isn't possible because it's variadic. I feel like the "right" way to handle this is to come up with some sort of "has side-effects" flag that lets a non-void operator be used as if it were void; I'm sure there are other operators where that could be convenient.

Reply:
Fair enough, that makes sense!